client/mapview_geometry.cpp

```diff
@@ -0,0 +1,189 @@
+ *
+ * SPDX-License-Identifier: GPLv3-or-later
+ */
+
+#include "mapview_geometry.h"
+
+#include "game.h"
+#include "map.h"
+#include "tilespec.h"
+
+namespace freeciv {
+
+/**
+ * @class gui_rect_iterator
+ * @brief Iterates over all map tiles that intersect with a rectangle in GUI
+ *        coordinates.
+ *
+ * The order of iteration is guaranteed to satisfy the painter's algorithm.
+ * The iteration covers not only tiles but tile edges and corners.
+ *
+ * Iteration is performed by repeatedly calling @ref next and querying the
+ * current @ref item_type "item type" with @ref current_item. If the item
+ * type matches your needs, you can retrieve the @ref corner, @ref edge, or
+ * @ref tile using the corresponding functions.
+ *
+ * Typical usage is as follows:
+ *
+ * ```{.cpp}
+ * for (auto it = gui_rect_iterator(tileset, rect); it.next(); ) {
+ *   if (it.current_item() == gui_rect_iterator::item_type::tile) {
+ *     auto tile = it.tile();
+ *     // Do something
+ *   }
+ * }
+ * ```
+ *
+ * One can also retrieve the canvas position of the current item using the
+ * @ref x and @ref y functions.
+ *
+ * @todo Recover the picture of the grid formerly at
+ *       http://article.gmane.org/gmane.games.freeciv.devel/50449.
+ */
+
+/**
+ * Constructor.
+ */
+gui_rect_iterator::gui_rect_iterator(const struct tileset *t,
+                                     const QRect &rect)
+{
+  const auto normalized = rect.normalized();
+  if (!normalized.isValid()) {
+    m_index = 0;
+    m_count = 0; // Don't iterate
+    return;
+  }
+
+  m_isometric = tileset_is_isometric(t);
+
+  m_r1 = m_isometric ? 2 : 1;
+  m_r2 = m_r1 * 2; // double the ratio
+  m_w = tileset_tile_width(t);
+  m_h = tileset_tile_height(t);
+
+  // Don't divide by r2 yet, to avoid integer rounding errors.
+  m_x0 = DIVIDE(normalized.left() * m_r2, m_w) - m_r1 / 2;
+  m_y0 = DIVIDE(normalized.top() * m_r2, m_h) - m_r1 / 2;
+  m_x1 = DIVIDE(normalized.right() * m_r2 + m_w - 1, m_w) + m_r1;
+  m_y1 = DIVIDE(normalized.bottom() * m_r2 + m_h - 1, m_h) + m_r1;
+
+  // Start counting
+  m_index = 0;
+  m_count = (m_x1 - m_x0) * (m_y1 - m_y0);
+
+  log_debug("Iterating over %d-%d x %d-%d rectangle.", m_x1, m_x0, m_y1,
+            m_y0);
+}
+
+/**
+ * Iterates to the next item. Returns `false` once the end of the iteration
+ * has been reached.
+ */
+bool gui_rect_iterator::next()
+{
+  if (m_index >= m_count) {
+    return false;
+  }
+
+  m_tile = nullptr;
+
+  m_xi = m_x0 + (m_index % (m_x1 - m_x0));
+  m_yi = m_y0 + (m_index / (m_x1 - m_x0));
+
+  const auto si = m_xi + m_yi;
+  const auto di = m_yi - m_xi;
+
+  if (m_isometric) {
+    if ((m_xi + m_yi) % 2 != 0) {
+      m_index++;
+      return next();
+    }
+
+    if (m_xi % 2 == 0 && m_yi % 2 == 0) {
+      if ((m_xi + m_yi) % 4 == 0) {
+        // Tile
+        m_type = item_type::tile;
+        m_tile = map_pos_to_tile(&(wld.map), si / 4 - 1, di / 4);
+      } else {
+        // Corner
```

Comment:
Take this chance to split this function into smaller functions ? I think some seperation here is good, like isometric and non isometric,  infact I'd split it into 4, iso corner, iso edge, non iso corner, non iso edge.

Reply:
Hmmm... it's quite a sensitive function and I ported it without really understanding how it works, basically just changing variable names. Splitting it would be even more risky.

Reply:
Alright, I'm merging and creating an issue for this.